flink-examples/flink-examples-streaming/src/main/scala/org/apache/flink/streaming/scala/examples/async/AsyncIOExample.scala

```diff
@@ -19,53 +19,69 @@
+import org.apache.flink.streaming.examples.async.util.SimpleSource
 
-import scala.concurrent.{ExecutionContext, Future}
+import java.util.concurrent.TimeUnit
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.util.{Failure, Success}
 
 object AsyncIOExample {
 
-  def main(args: Array[String]) {
-    val timeout = 10000L
-
-    val env = StreamExecutionEnvironment.getExecutionEnvironment
+  /** An example of a [[RichAsyncFunction]] using an async client to query an external service. */
+  class SampleAsyncFunction extends RichAsyncFunction[Int, String] {
+    private var client: AsyncClient = _
 
-    val input = env.addSource(new SimpleSource())
+    override def open(parameters: Configuration): Unit = {
+      client = new AsyncClient
+    }
 
-    val asyncMapped = AsyncDataStream.orderedWait(input, timeout, TimeUnit.MILLISECONDS, 10) {
-      (input, collector: ResultFuture[Int]) =>
-        Future {
-          collector.complete(Seq(input))
-        } (ExecutionContext.global)
+    override def asyncInvoke(input: Int, resultFuture: ResultFuture[String]): Unit = {
+      client.query(input).onComplete {
+        case Success(value) => resultFuture.complete(Seq(value))
+        case Failure(exception) => resultFuture.completeExceptionally(exception)
+      }
     }
+  }
 
-    asyncMapped.print()
+  def main(args: Array[String]): Unit = {
+    val params = ParameterTool.fromArgs(args)
 
-    env.execute("Async I/O job")
-  }
-}
+    var mode: String = null
+    var timeout = 0L
 
-class SimpleSource extends ParallelSourceFunction[Int] {
-  var running = true
-  var counter = 0
+    try {
+      mode = params.get("waitMode", "ordered")
+      timeout = params.getLong("timeout", 10000L)
+    } catch {
+      case e: Exception =>
+        println("To customize example, use: AsyncIOExample [--waitMode <ordered or unordered>]")
+        throw e
+    }
 
-  override def run(ctx: SourceContext[Int]): Unit = {
-    while (running) {
-      ctx.getCheckpointLock.synchronized {
-        ctx.collect(counter)
-      }
-      counter += 1
+    // obtain execution environment
+    val env = StreamExecutionEnvironment.getExecutionEnvironment
 
-      Thread.sleep(10L)
+    // create input stream of a single integer
+    val inputStream = env.addSource(new SimpleSource).map(_.toInt)
+
+    val function = new SampleAsyncFunction
+
+    // add async operator to streaming job
+    val result = mode.toUpperCase match {
+      case "ORDERED" =>
+        AsyncDataStream.orderedWait(inputStream, function, timeout, TimeUnit.MILLISECONDS, 20)
+      case "UNORDERED" =>
+        AsyncDataStream.unorderedWait(inputStream, function, timeout, TimeUnit.MILLISECONDS, 20)
+      case _ => throw new IllegalStateException("Unknown mode: " + mode)
     }
-  }
 
-  override def cancel(): Unit = {
-    running = false
+    result.print()
+
+    // execute the program
+    env.execute("Async IO Example")
```

Comment:
What about appending the mode information of the async operator into the job name?

Reply:
Great idea!